app/models/user.rb

```diff
@@ -418,6 +422,49 @@ def setup(opts)
     self
   end
 
+  def pam_managed_user?
+    return false unless Devise.pam_authentication
+    name = pam_get_name
+    return false unless name
+    Rpam2.account(Devise.pam_default_service, name).present?
+  end
+
+  def no_pam_conflict
+    errors.add(:username, "username conflicts with pam account") if pam_conflict?
+  end
+
+  def pam_conflict?
+    encrypted_password.present? && pam_managed_user?
+  end
+
+  def pam_conflict(_attributes)
+    # block pam logins on users with pw
+    nil
+  end
+
+  def pam_setup(attributes)
+    args = {}
+    args['username'] = self[::Devise.usernamefield]
+    args['email'] = Rpam2.getenv(find_pam_service, self[::Devise.usernamefield], attributes[:password], 'email')
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.